File: y2018/python/day17/day17.py
import re
import numpy as np
from water import Tile, TILE_STR  # TILE_STR is for debugging

# like always, positions are in (row, col) format
WATER = (0, 500)

xy_fmt = r'x=(\d+),\s+y=(\d+)\.\.(\d+)'
yx_fmt = r'y=(\d+),\s+x=(\d+)\.\.(\d+)'
clay = set()
with open('day17.txt') as read:
    for c in read.readlines():
        if re.match(xy_fmt, c) is not None:
            x, ys, ye = map(int, next(iter(re.findall(xy_fmt, c))))
            for y in range(ys, ye + 1):
                clay.add((y, x))
        elif re.match(yx_fmt, c) is not None:
            y, xs, xe = map(int, next(iter(re.findall(yx_fmt, c))))
            for x in range(xs, xe + 1):
                clay.add((y, x))

min_r = WATER[0]
max_r = WATER[0]
min_c = WATER[1]
max_c = WATER[1]
for r, c in clay:
    min_r = min(min_r, r)
    max_r = max(max_r, r)
    min_c = min(min_c, c)
    max_c = max(max_c, c)

"""
the extra 1 row is so i don't have to do bounds checking
the extra 2 columns are because water can flow right on the edge of the grid
"""
grid = np.empty((max_r - min_r + 2, max_c - min_c + 3), dtype=Tile)
grid.fill(Tile.SAND)
for r, c in clay:
    grid[r - min_r, c - min_c + 1] = Tile.CLAY

# algorithm sauce: https://www.michaelfogleman.com/aoc18/#17
blocking = [Tile.CLAY, Tile.STILL]
# position, (true = falling water, false = spreading water)
todo = [(WATER[0] - min_r, WATER[1] - min_c + 1, True)]
seen = set()
while todo:
    r, c, type_ = todo.pop()
    if (r, c, type_) in seen:
        continue
    seen.add((r, c, type_))
    if type_:
        while r <= max_r and grid[r + 1, c] not in blocking:
            grid[r, c] = Tile.FLOW
            r += 1
        if r <= max_r:
            grid[r, c] = Tile.FLOW
            todo.append((r, c, False))
    else:
        c_left = c
        while (grid[r + 1, c_left] in blocking
               and grid[r, c_left - 1] != Tile.CLAY):
            c_left -= 1

        c_right = c
        while (grid[r + 1, c_right] in blocking
               and grid[r, c_right + 1] != Tile.CLAY):
            c_right += 1

        left_wall = grid[r, c_left - 1] == Tile.CLAY
        right_wall = grid[r, c_right + 1] == Tile.CLAY
        if left_wall and right_wall:
            for fill_c in range(c_left, c_right + 1):
                grid[r, fill_c] = Tile.STILL
            if r != 0:
                todo.append((r - 1, c, False))
        else:
            for fill_c in range(c_left, c_right + 1):
                grid[r, fill_c] = Tile.FLOW
            if not left_wall:
                todo.append((r, c_left, True))
            if not right_wall:
                todo.append((r, c_right, True))

highest_clay = np.where(grid == Tile.CLAY)[0][0]
flowing = (grid == Tile.FLOW).sum() - highest_clay
still = (grid == Tile.STILL).sum()
print(f"total water (while flowing): {flowing + still}")
print(f"total water (after flowing stops): {still}")

File: y2018/rust/src/bin/day17.rs
use regex::Regex;
use std::collections::HashSet;
use std::fs;

const WATER: (usize, usize) = (0, 500);

#[derive(Copy, Clone, Eq, PartialEq)]
enum Tile {
    Clay,
    Sand,
    Still,
    Flow,
}

fn main() {
    let xy_fmt = Regex::new(r"x=(\d+),\s+y=(\d+)\.\.(\d+)").unwrap();
    let yx_fmt = Regex::new(r"y=(\d+),\s+x=(\d+)\.\.(\d+)").unwrap();
    let mut clay = HashSet::new();
    for line in fs::read_to_string("../input/day17.txt").unwrap().lines() {
        if let Some(m) = xy_fmt.captures(line) {
            let (x, ys, ye) = (
                m[1].parse::<usize>().unwrap(),
                m[2].parse::<usize>().unwrap(),
                m[3].parse::<usize>().unwrap(),
            );
            for y in ys..=ye {
                clay.insert((y, x));
            }
        } else if let Some(m) = yx_fmt.captures(line) {
            let (y, xs, xe) = (
                m[1].parse::<usize>().unwrap(),
                m[2].parse::<usize>().unwrap(),
                m[3].parse::<usize>().unwrap(),
            );
            for x in xs..=xe {
                clay.insert((y, x));
            }
        }
    }

    // all of this is from the python version, i'm just redoing it in rust here
    let mut max_r = WATER.0;
    let mut min_c = WATER.1;
    let mut max_c = WATER.1;
    clay.iter().for_each(|(y, x)| {
        max_r = max_r.max(*y);
        min_c = min_c.min(*x);
        max_c = max_c.max(*x);
    });

    let mut grid = vec![vec![Tile::Sand; max_c - min_c + 3]; max_r + 2];
    clay.iter()
        .for_each(|c| grid[c.0][c.1 - min_c + 1] = Tile::Clay);

    let blocking = vec![Tile::Clay, Tile::Still];
    let mut todo = vec![(WATER.0, WATER.1 - min_c + 1, true)];
    let mut seen = HashSet::new();
    while !todo.is_empty() {
        let curr = todo.pop().unwrap();
        if seen.contains(&curr) {
            continue;
        }
        seen.insert(curr);
        let (mut r, c, type_) = curr;
        if type_ {
            while r <= max_r && !blocking.contains(&grid[r + 1][c]) {
                grid[r][c] = Tile::Flow;
                r += 1;
            }
            if r <= max_r {
                grid[r][c] = Tile::Flow;
                todo.push((r, c, false));
            }
        } else {
            let mut c_left = c;
            while blocking.contains(&grid[r + 1][c_left]) && grid[r][c_left - 1] != Tile::Clay {
                c_left -= 1;
            }

            let mut c_right = c;
            while blocking.contains(&grid[r + 1][c_right]) && grid[r][c_right + 1] != Tile::Clay {
                c_right += 1;
            }

            let left_wall = c_left > 0 && grid[r][c_left - 1] == Tile::Clay;
            let right_wall = c_right + 1 < grid[r].len() && grid[r][c_right + 1] == Tile::Clay;
            if left_wall && right_wall {
                for fill_c in c_left..=c_right {
                    grid[r][fill_c] = Tile::Still;
                }
                if r != 0 {
                    todo.push((r - 1, c, false));
                }
            } else {
                for fill_c in c_left..=c_right {
                    grid[r][fill_c] = Tile::Flow;
                }
                if !left_wall {
                    todo.push((r, c_left, true));
                }
                if !right_wall {
                    todo.push((r, c_right, true));
                }
            }
        }
    }

    let highest_clay = clay.iter().min_by_key(|c| c.0).unwrap().0 as u32;
    let flowing = grid
        .iter()
        .flat_map(|r| r)
        .map(|t| (*t == Tile::Flow) as u32)
        .sum::<u32>()
        - highest_clay;
    let still = grid
        .iter()
        .flat_map(|r| r)
        .map(|t| (*t == Tile::Still) as u32)
        .sum::<u32>();

    println!("total water (while flowing): {}", flowing + still);
    println!("total water (while still): {still}");
}
